services/github/github-issues-search.service.js

```diff
@@ -0,0 +1,112 @@
+'use strict'
+
+const gql = require('graphql-tag')
+const Joi = require('joi')
+const { metric } = require('../text-formatters')
+const { nonNegativeInteger } = require('../validators')
+const { GithubAuthV4Service } = require('./github-auth-service')
+const { documentation, transformErrors } = require('./github-helpers')
+
+const issueCountSchema = Joi.object({
+  data: Joi.object({
+    search: Joi.object({
+      issueCount: nonNegativeInteger,
+    }).required(),
+  }).required(),
+}).required()
+
+const queryParamSchema = Joi.object({
+  query: Joi.string().required(),
+}).required()
+
+class BaseGithubIssuesSearch extends GithubAuthV4Service {
+  static category = 'issue-tracking'
+  static defaultBadgeData = { label: 'query', color: 'informational' }
+
+  static render({ issueCount }) {
+    return { message: metric(issueCount) }
+  }
+
+  async fetch({ query }) {
+    const data = await this._requestGraphql({
+      query: gql`
+        query($query: String!) {
+          search(query: $query, type: ISSUE) {
+            issueCount
+          }
+        }
+      `,
+      variables: { query },
+      schema: issueCountSchema,
+      transformErrors,
+    })
+    return data.data.search.issueCount
+  }
+}
+
+class GithubIssuesSearch extends BaseGithubIssuesSearch {
+  static route = {
+    base: 'github',
+    pattern: 'issues-search',
+    queryParamSchema,
+  }
+
+  static examples = [
+    {
+      title: 'GitHub issue custom search',
+      namedParams: {},
+      queryParams: {
+        query: 'repo:badges/shields is:closed label:bug author:app/sentry-io',
+      },
+      staticPreview: {
+        label: 'query',
+        message: '10',
+        color: 'blue',
+      },
+      documentation,
+    },
+  ]
+
+  async handle(namedParams, { query }) {
+    const issueCount = await this.fetch({ query })
```

Comment:
I don't recall the set of allowed/used characters in a query but any risk of uri encoding/decoding issues with the query?

Reply:
You need to URL encode your query appropriately in the badge URL, but by the time we get to `handle()` it has been decoded elsewhere i.e: if you call `/issues-search/badges/shields?query=is%3Aclosed%20label%3Abug%20author%3Aapp%2Fsentry-io` when we invoke `handle()` the value of `query` is `is:closed label:bug author:app/sentry-io` not `is%3Aclosed%20label%3Abug%20author%3Aapp%2Fsentry-io` so there's nothing extra to do there. If you don't URLEncode the quotes and colons and spaces and things in the URL it won't work but y'know.. that's URLs for you :shrug: nothing special going on here.

Reply:
The other type of escaping worth commenting on is when we pass the query in the GraphQL Body.

Because the query is a GraphQL variable (we're not using string concatenation), we also don't need to escape the quotes, so whereas in https://github.com/badges/shields/discussions/5943#discussioncomment-211217 we need to escape the quotes, with this badge if the query is `repo:TheIOFoundation/ProjectLockdown is:open is:issue label:"Need: Position" label:"Stage: Ready"` we can call `/issues-search?query=repo%3ATheIOFoundation%2FProjectLockdown%20is%3Aopen%20is%3Aissue%20label%3A%22Need%3A%20Position%22%20label%3A%22Stage%3A%20Ready%22` and neither the user nor us need to directly escape the quote characters in that query.